### PR TITLE
Fix issue 20625: Improve diagnostics and test immediately called function literals

### PR DESCRIPTION
**Pull request by Geod24**
The didn't seem to be any test in the testsuite covering this.
Additionally, the error message was just awful,
and our users are used to so much better nowadays.

### CODE REVIEW

**Review by thewilsonator — APPROVED** [review]

**Comment by thewilsonator** [comment]
Not that it particularly needs it, but do you want to reference an issue or add a changelog entry for this?

**Comment by Geod24** [comment]
Added a bugzilla entry for it.

**Comment by thewilsonator** [comment]
Still need to retitle the commit to fix the issue.

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -4056,32 +4063,41 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
-                for (size_t i = 0; i < exp.td.parameters.dim; i++)
+            for (size_t i = 0; i < exp.td.parameters.dim; i++)
+            {
+                TemplateParameter tp = (*exp.td.parameters)[i];
+                for (size_t u = 0; u < dim; u++)
                 {
-                    TemplateParameter tp = (*exp.td.parameters)[i];
-                    for (size_t u = 0; u < dim; u++)
+                    Parameter p = tfl.parameterList[u];
+                    if (p.type.ty == Tident && (cast(TypeIdentifier)p.type).ident == tp.ident)
```
Comment (WalterBright) [inline src/dmd/expressionsem.d]: Use `isTypeIdentifier()` instead

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -4056,32 +4063,41 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
             {
-                auto tiargs = new Objects();
-                tiargs.reserve(exp.td.parameters.dim);
+                OutBuffer buf;
+                foreach (idx, ref arg; *arguments)
+                    buf.printf("%s%s", (idx ? ", ".ptr : "".ptr), arg.type.toChars());
+                exp.error("function literal `%s%s` is not callable using argument types `(%s)`",
+                          exp.fd.toChars(), parametersTypeToChars(tfl.parameterList),
+                          buf.peekChars());
+                exp.errorSupplemental("too %s arguments, expected `%d`, got `%d`",
```
Comment (WalterBright) [inline src/dmd/expressionsem.d]: Use either `%zd` for `size_t` or cast the values to `int`.

Comment (WalterBright) [reply src/dmd/expressionsem.d]: Prolly the cast because some old VC++'s don't support `z`.

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (Geod24) [reply src/dmd/expressionsem.d]: I didn't touch that part of the code, just removed indentation.
Enable `Hide whitespace change` to see what's actually going on.
I try not to refactor too much when fixing an issue, as per your guidelines.

File: src/dmd/expressionsem.d (hunk 2 above)
Comment (Geod24) [reply src/dmd/expressionsem.d]: Done

**Comment by Geod24** [comment]
> Still need to retitle the commit to fix the issue.

The ol' "Forgot to push". Fixed now (and addressed printf's comments)

**Comment by Geod24** [comment]
@WalterBright : This is my time to remind you that we could really benefit from having a more clever interface than `printf`-like functionality.
I made an attempt in that direction a few years ago, and while it was not perfect, it would have avoided many of the `cast(int)` you had to add when adding `printf` checks: https://github.com/dlang/dmd/pull/8687 .

There are other alternatives to consider... For example, I find it annoying that one has to use `printf`-like format specifier in a language that doesn't have `printf` limitations (read: a language that can know about the type of variadic arguments).

Having worked with Tango's approach (borrowed from C#), I miss writing :
```D
format("expected {} elements, got {}", dim, arguments.dim);
```
and just letting the language do its job.  There's even [a pretty well documented implementation](https://github.com/sociomantic-tsunami/ocean/blob/b0b0a82c08148cf1382b01764b0f0961641c78e6/src/ocean/text/convert/Formatter.d#L16-L44) out there.

While that boat has sailed for Phobos, it hasn't for DMD, as we are committed in not using Phobos.

**Pull request merged**